src/libsyntax/parse/parser.rs

```diff
@@ -233,6 +233,21 @@ macro_rules! maybe_whole {
     )
 }
 
+/// Uses $parse_expr to parse an expression and returns the span of the interpolated
+/// token or the span of the parsed expression, if it was not interpolated
+macro_rules! interpolated_or_expr_span {
```

Comment:
I now remembered to original reason for that being a macro. As it was, `interpolated_or_expr_span` used the current token to check if it is interpolated, which means the parse functions has to be called after that check.

But by using `last_token` it works as expected. I pushed another commit however, which avoids storing (and cloning) interpolated tokens and instead added a `Parser.last_token_interpolated` flag.